source/validate_cfg.cpp

```diff
@@ -24,32 +24,314 @@
+  struct block_info {
+    cbb_ptr block;
+    bb_iter current_successor;
+    bb_iter end_successor;
+  };
+
+  vector<cbb_ptr> out;
+  vector<block_info> staged;
+  vector<uint32_t> processed;
+
+  staged.reserve(size);
+  staged.emplace_back(block_info{&entry, begin(entry.get_successors()),
+                                 end(entry.get_successors())});
+  processed.push_back(entry.get_id());
+
+  while (!staged.empty()) {
+    block_info& top = staged.back();
+    if (top.current_successor == top.end_successor) {
+      // No children left to process
+      out.push_back(top.block);
+      staged.pop_back();
+    } else {
+      bb_iter& child_iter = top.current_successor;
+
+      if (!Contains(processed, (*child_iter)->get_id())) {
+        // Process next child
+        staged.emplace_back(block_info{*child_iter,
+                                       begin((*child_iter)->get_successors()),
+                                       end((*child_iter)->get_successors())});
+        processed.push_back((*child_iter)->get_id());
+      }
+      child_iter++;
+    }
+  }
+  return out;
+}
+}
+
+vector<pair<BasicBlock*, BasicBlock*>> CalculateDominators(
+    const BasicBlock& first_block) {
+  struct block_detail {
+    size_t dominator;
+    size_t postorder_index;
+  };
+
+  vector<cbb_ptr> postorder = PostOrderSort(first_block, 10);
+  const size_t undefined_dom = static_cast<uint32_t>(postorder.size());
+
+  unordered_map<cbb_ptr, block_detail> idoms;
+  for (size_t i = 0; i < postorder.size(); i++) {
+    idoms[postorder[i]] = {undefined_dom, i};
+  }
+
+  idoms[postorder.back()].dominator = idoms[postorder.back()].postorder_index;
+
+  bool changed = true;
+  while (changed) {
+    changed = false;
+    for (auto b = postorder.rbegin() + 1; b != postorder.rend(); b++) {
+      // printf("processing: %d\n", (*b)->get_id());
+      size_t& b_dom = idoms[*b].dominator;
+      const vector<BasicBlock*>& predecessors = (*b)->get_predecessors();
+
+      // first processed predecessor
+      BasicBlock* idom =
+          *find_if(begin(predecessors), end(predecessors),
+                   [&idoms, undefined_dom](BasicBlock* pred) {
+                     return idoms[pred].dominator != undefined_dom;
+                   });
+      size_t idom_idx = idoms[idom].postorder_index;
+
+      // all other predecessors
+      for (auto p : predecessors) {
+        if (idom == p) {
+          continue;
+        }
+        if (idoms[p].dominator != undefined_dom) {
+          size_t finger1 = idoms[p].postorder_index;
+          size_t finger2 = idom_idx;
+          while (finger1 != finger2) {
+            while (finger1 < finger2) {
+              finger1 = idoms[postorder[finger1]].dominator;
+            }
+            while (finger2 < finger1) {
+              finger2 = idoms[postorder[finger2]].dominator;
+            }
+          }
+          idom_idx = finger1;
+        }
+      }
+      if (b_dom != idom_idx) {
+        b_dom = idom_idx;
```

Comment:
This line updates the reference to idoms[*b].dominator. it is not an assignment to the local variable.